lib/configurators/databases/aws.go

```diff
@@ -0,0 +1,656 @@
+	}
+
+	a.formattedPolicyBoundaryDocument, err = formatPolicyDocument(a.policyBoundaryDocument)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// retrieveCurrentPolicy returns information from AWS about a policy. If the
+// policy doesn't have `teleport` tag and error will be returned.
+func (a *awsConfigurator) retrieveCurrentPolicy(policyName string) (string, []*iam.PolicyVersion, error) {
+	if a.awsIdentity == nil {
+		return "", nil, trace.NotFound("AWS identity not provided, unable to determine policy account ID")
+	}
+
+	policyArn := fmt.Sprintf("arn:aws:iam::%s:policy/%s", a.awsIdentity.GetAccountID(), policyName)
+	client, err := a.getAWSIAMClient()
+	if err != nil {
+		return "", nil, trace.Wrap(err)
+	}
+
+	getPolicyResp, err := client.GetPolicy(&iam.GetPolicyInput{PolicyArn: aws.String(policyArn)})
+	if err != nil {
+		return policyArn, nil, wrapAWSError(err)
+	}
+
+	// Check if the policy has Teleport tag.
+	if !hasTeleportTag(getPolicyResp.Policy.Tags) {
+		return policyArn, nil, trace.BadParameter("Policy %q already exists and was not created by Teleport configurator.", policyName)
+	}
+
+	resp, err := client.ListPolicyVersions(&iam.ListPolicyVersionsInput{PolicyArn: aws.String(policyArn)})
+	if err != nil {
+		return policyArn, nil, wrapAWSError(err)
+	}
+
+	return policyArn, resp.Versions, nil
+}
+
+// upsertPolicyDocument creates a new Policy or creates a Policy version if a
+// policy with the same name already exists.
+//
+// Since policies can hold only five versions, we need to delete a policy
+// version (if the limit is reached) and create a new version.
+// Ref: https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_managed-versioning.html#version-limits
+func (a *awsConfigurator) upsertPolicyDocument(documentName, documentDescription string, document *awslib.PolicyDocument) (string, error) {
+	encodedPolicyDocument, err := encodePolicyDocument(document)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	iamClient, err := a.getAWSIAMClient()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	//
+	policyArn, versions, err := a.retrieveCurrentPolicy(documentName)
+	if err != nil && !trace.IsNotFound(err) {
+		return "", trace.Wrap(err)
+	}
+
+	// If no versions were found, we need to create a new policy.
+	if trace.IsNotFound(err) {
+		resp, err := iamClient.CreatePolicyWithContext(context.Background(), &iam.CreatePolicyInput{
+			Description:    aws.String(documentDescription),
+			PolicyDocument: aws.String(encodedPolicyDocument),
+			PolicyName:     aws.String(documentName),
+			Tags: []*iam.Tag{
+				{Key: aws.String(policyTeleportTagKey), Value: aws.String(policyTeleportTagValue)},
+			},
+		})
+		if err != nil {
+			return "", wrapAWSError(err)
+		}
+
+		return *resp.Policy.Arn, nil
+	}
+
+	// Check number of policy versions and delete one if necessary.
+	if len(versions) == 5 {
+		// Sort versions based on create date.
+		sort.Slice(versions, func(i, j int) bool {
```

Comment:
Can this logic be factored out into `DeleteOldestPolicy` or something similar e.g. on that AWSClient wrapper I suggested above.